lua/lean/rpc.lua

```diff
@@ -38,40 +46,81 @@ function Session:new(bufnr, uri)
 end
 
 function Session:close()
+  self:release_now{}
   self.keepalive_timer:close()
-end
-
----@param session_id string
-function Session:connected(session_id)
-  self.session_id = session_id
-  self.on_connected:notify_all()
+  self.closed = true
 end
 
 ---@param refs RpcRef[]
-function Session:release(refs)
-  if self.closed or #refs == 0 then return end
+function Session:release_now(refs)
+  for _, ptr in ipairs(refs) do table.insert(self.to_release, ptr) end
+  if self.closed or #self.to_release == 0 then return end
+  print('releasing ' .. #self.to_release)
   vim.lsp.buf_notify(self.bufnr, '$/lean/rpc/release', {
     uri = self.uri,
     sessionId = self.session_id,
-    refs = refs,
+    refs = self.to_release,
   })
+  self.to_release = {}
 end
 
+---@param refs RpcRef[]
+function Session:release_deferred(refs)
+  for _, ptr in ipairs(refs) do table.insert(self.to_release, ptr) end
+  if self.release_timer == nil then
+    self.release_timer = vim.defer_fn(function ()
+      self.release_timer = nil
+      self:release_now{}
+    end, 100)
+  end
+end
+
+---@class TextDocumentIdentifier
+
+---@class LspPosition
+---@field line integer
+---@field character integer
+
+---@class LspRange
+---@field start LspPosition
+---@field end LspPosition
+
 ---@class TextDocumentPositionParams
+---@field textDocument TextDocumentIdentifier
+---@field position LspPosition
 
 ---@param pos TextDocumentPositionParams
 ---@param method string
 ---@return any result
 ---@return any error
 function Session:call(pos, method, params)
-  if self.session_id == nil then
+  if not self.connected then
```

Comment:
Would there be any harm in making this a `while`?  Just in case it disconnects for some reason in one of the calls.